embed/trezorhal/image.c

```diff
@@ -29,27 +29,29 @@ static bool compute_pubkey(uint8_t sig_m, uint8_t sig_n, const uint8_t * const *
     return 0 == ed25519_cosi_combine_publickeys(res, keys, sig_m);
 }
 
-bool image_parse_header(const uint8_t *data, uint32_t magic, uint32_t maxsize, image_header *hdr)
+bool image_parse_header(const uint8_t * const data, const uint32_t magic, const uint32_t maxsize, image_header * const hdr)
 {
-    memcpy(&hdr->magic, data, 4);
+    uint32_t *data_ptr = (uint32_t *) data;
+
+    hdr->magic = *data_ptr++;
     if (hdr->magic != magic) return false;
 
-    memcpy(&hdr->hdrlen, data + 4, 4);
+    hdr->hdrlen = *data_ptr++;
     if (hdr->hdrlen != HEADER_SIZE) return false;
 
-    memcpy(&hdr->expiry, data + 8, 4);
-    if (hdr->expiry != 0) return false;
+    hdr->expiry = *data_ptr++;
+    if (hdr->expiry != 0) return false; // todo: then what is the point of having this field?
 
-    memcpy(&hdr->codelen, data + 12, 4);
-    if (hdr->hdrlen + hdr->codelen < 4 * 1024) return false;
-    if (hdr->hdrlen + hdr->codelen > maxsize) return false;
-    if ((hdr->hdrlen + hdr->codelen) % 512 != 0) return false;
+    hdr->codelen = *data_ptr++;
+    if ((hdr->hdrlen + hdr->codelen) < 4 * 1024) return false;
+    if ((hdr->hdrlen + hdr->codelen) > maxsize) return false;
+    if (((hdr->hdrlen + hdr->codelen) & 511) != 0) return false;
 
-    memcpy(&hdr->version, data + 16, 4);
+    hdr->version = *data_ptr++;
 
     // uint8_t reserved[427];
 
-    memcpy(&hdr->sigmask, data + 0x01BF, 1);
+    hdr->sigmask = *(data + 0x01BF);
```

Comment:
I was using memcpy instead of assignement to avoid unaligned access, which we had problem with in the past. Are you sure this will not cause some issues? (The question is valid for all memcpy called replaced with assignments, not just this one).

Reply:
interesting concern. i think the pointer is ok because the data pointer is aligned in current uses, and the cortex-m4 processor supports unaligned access for the LDR instruction (i tried an unaligned access and it did not blow up). but to be safe, i'll revert these changes back to memcpy to be extra safe.